File: 0104-maximum-depth-of-binary-tree/0104-maximum-depth-of-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

from collections import deque
class Solution:
    def maxDepth(self, root: Optional[TreeNode]) -> int:
        
        if not root:
            return 0

        stack = [(root, 1)]
        maxDepth = 0

        while stack:
            node, depth = stack.pop()
            
            #Calculate maxDepth
            maxDepth = max(maxDepth, depth)

            if node.left:
                stack.append((node.left, depth + 1))
            if node.right:
                stack.append((node.right, depth + 1))
        
        return maxDepth

File: 0104-maximum-depth-of-binary-tree/0104-maximum-depth-of-binary-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
#include <stack> 
class Solution {
public:
    int maxDepth(TreeNode* root) {
        
        if (!root)
        {
            return 0;
        }

        stack<pair<TreeNode*, int>> stack;
        stack.push({root, 1});

        while (!stack.empty())
        {

            auto [node, count] = stack.top();
            stack.pop();

            if (node->left)
            {
                stack.push({node->left, count + 1});
            }

            if (node->right)
            {
                stack.push({node->right, count + 1});
            }

            if (count > res)
            {
                res = count;
            }
        }

        return res;

    }
private:
    int res = 0;
};